backend/azure/dtmb/dtmb.go

```diff
@@ -0,0 +1,531 @@
+package dtmb
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/microsoft/durabletask-go/api"
+	"github.com/microsoft/durabletask-go/backend"
+	"github.com/microsoft/durabletask-go/internal/protos"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	dtmbprotos "github.com/microsoft/durabletask-go/backend/azure/dtmb/internal/backend/v1"
+	"github.com/microsoft/durabletask-go/backend/azure/dtmb/internal/utils"
+)
+
+const (
+	// DefaultEndpoint is the default endpoint for the DTMB service.
+	defaultEndpoint = "localhost:50051"
+)
+
+type dtmb struct {
+	logger                    backend.Logger
+	endpoint                  string
+	options                   *DTMBOptions
+	orchestrationQueue        utils.SyncQueue[dtmbprotos.ExecuteOrchestrationMessage]
+	activityQueue             utils.SyncQueue[dtmbprotos.ExecuteActivityMessage]
+	orchestrationHistoryCache utils.OrchestrationHistoryCache
+	connectWorkerClientStream chan *dtmbprotos.ConnectWorkerClientMessage
+	clientClient              dtmbprotos.TaskHubClientClient
+	workerClient              dtmbprotos.TaskHubWorkerClient
+	workerCancelFunc          context.CancelFunc
+}
+
+type DTMBOptions struct {
+	Endpoint string
+}
+
+func NewDTMBOptions(endpoint string) *DTMBOptions {
+	if endpoint != "" {
+		return &DTMBOptions{
+			Endpoint: endpoint,
+		}
+	} else {
+		return &DTMBOptions{
+			Endpoint: defaultEndpoint,
+		}
+	}
+}
+
+func NewDTMB(opts *DTMBOptions, logger backend.Logger) (*dtmb, error) {
+	be := &dtmb{
+		logger: logger,
+	}
+
+	if opts == nil {
+		opts = NewDTMBOptions(defaultEndpoint)
+	}
+	be.options = opts
+	be.endpoint = opts.Endpoint
+
+	// The following queues are used to store messages received from the server
+	be.orchestrationQueue = utils.NewSyncQueue[dtmbprotos.ExecuteOrchestrationMessage]()
+	be.activityQueue = utils.NewSyncQueue[dtmbprotos.ExecuteActivityMessage]()
+
+	be.orchestrationHistoryCache = utils.NewOrchestrationHistoryCache(nil) // TODO: make capacity configurable
+
+	ctx := context.Background()
+	connCtx, connCancel := context.WithTimeout(ctx, 15*time.Second) // TODO: make this a configurable timeout
+	conn, err := grpc.DialContext(connCtx, be.endpoint, grpc.WithBlock(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	connCancel()
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to dtmb: %v", err)
+	}
+
+	be.clientClient = dtmbprotos.NewTaskHubClientClient(conn)
+	be.workerClient = dtmbprotos.NewTaskHubWorkerClient(conn)
+
+	return be, nil
+}
+
+// CreateTaskHub creates a new task hub for the current backend. Task hub creation must be idempotent.
+//
+// If the task hub for this backend already exists, an error of type [ErrTaskHubExists] is returned.
+func (d *dtmb) CreateTaskHub(ctx context.Context) error {
+	_, err := d.clientClient.Metadata(ctx, &dtmbprotos.MetadataRequest{})
```

Comment:
Add a comment explaining that on DTMB the Task Hub must be created on the service, and here we just check if it exists